expression/builtin_time.go

```diff
@@ -5824,16 +5824,20 @@ func (c *makeTimeFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		return nil, err
 	}
 	tp, decimal := args[2].GetType().EvalType(), 0
+
 	switch tp {
+	case types.ETString:
+		decimal = 6
 	case types.ETInt:
-	case types.ETReal, types.ETDecimal:
-		decimal = args[2].GetType().Decimal
-		if decimal > 6 || decimal == types.UnspecifiedLength {
-			decimal = 6
-		}
+		decimal = 0
 	default:
+		decimal = args[2].GetType().Decimal
```

Comment:
Can we use this to handle all the cases?

Reply:
I add mores unit to cover all the cases in ` expression/typeinfer_test.go`. It seems to be good.

Reply:
Can we set the decimal be `UnspecifiedLength`, if the type is `ETSTRING` ?

Reply:
For ETString, the decimal shoule be 31? Then it will be truncated to 6,

Reply:
I think set `UnspecifiedLength` for `ETSTRING` is more clear, because decimal is 31 in MySQL but -1 in TiDB.Decimal use 31 in `ETSTRING` is not same as common rules in TIDB.

Reply:
Anyway, it's covered by L5836. We can set decimal = args[2].GetType().Decimal for ETString.

Reply:
Set decimal = args[2].GetType().Decimal for ETString can't cover the situation that the arg is varchar type(`{maketime(c_int_d, c_int_d, c_char)`  test in `expression/typeinfer_test.go`). 
In that case, the decimal is zero not the UnspecifiedLength, and we will get zero decimal and it is wrong (expect decimal is 6).
So we have to specialize UnspecifiedLength in  decimal for type ETSTRING.

Reply:
For literal "string", the decimal is 31.
```
mysql> select "asd";
Field   1:  `asd`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_bin (46)
Length:     12
Max_length: 3
Decimals:   31
Flags:      NOT_NULL


+-----+
| asd |
+-----+
| asd |
+-----+
1 row in set (0.01 sec)
```
Could you find out why it's 0 for the string column?